### PR TITLE
Kr-BoxClusterFinder can now take arbitrary amounts of timebins

### PR DESCRIPTION
**Pull request by HauenderPhilip**
Instead of a fixed 3D map, a std::deque is now used.
See comments in code.

Still to do: Re-implement functions for workflow.

### CODE REVIEW

File: Detectors/TPC/reconstruction/src/KrBoxClusterFinder.cxx [hunk 1]
```diff
@@ -112,7 +86,53 @@ void KrBoxClusterFinder::fillADCValue(int cru, int rowInSector, int padInRow, in
...
+}
+
+void KrBoxClusterFinder::resetADCMap() {
+  // Has to be reimplemented!
+  return;
+}
+
+void KrBoxClusterFinder::fillADCValue(int cru, int rowInSector, int padInRow, int timeBin, float adcValue) {
+  // Has to be reimplemented!
+  return;
```
Comment (HauenderPhilip) [inline Detectors/TPC/reconstruction/src/KrBoxClusterFinder.cxx]: These functions are needed for the workflow but do not work anymore as they were implemented.

File: Detectors/TPC/reconstruction/include/TPCReconstruction/KrBoxClusterFinder.h [hunk 2]
```diff
@@ -206,16 +222,36 @@ class KrBoxClusterFinder
 
   KrCluster mTempCluster; ///< Used to save the cluster data
 
-  /// Here the map is defined where all digits are temporarily stored
-  std::array<std::array<std::array<float, MaxPads>, MaxRows>, MaxTimes> mMapOfAllDigits{};
+  /// A temporary timeslice that gets filled and can be added to the set of timeslices
+  std::array<std::array<float, MaxPads>, MaxRows> mTempTimeSlice;
```
Comment (wiechula) [inline Detectors/TPC/reconstruction/include/TPCReconstruction/KrBoxClusterFinder.h]:
```suggestion
  using TimeSliceSector = std::array<std::array<float, MaxPads>, MaxRows>;
  /// A temporary timeslice that gets filled and can be added to the set of timeslices
  TimeSliceSector mTempTimeSlice;
```

File: Detectors/TPC/reconstruction/include/TPCReconstruction/KrBoxClusterFinder.h [hunk 3]
```diff
@@ -206,16 +222,36 @@ class KrBoxClusterFinder
...
+  ///      .............
+  /// n-1  o o o x o o o
+  ///   n  o o o x o o o
+  ///
+  /// \endverbatim
+  /// 
+  /// x-axis: Timeslice number
+  /// y-axis: Pad number
+  /// Time slice four is the interesting one. In there, local maxima are found and clusters are built from it. After it is processed, timeslice number 1 will be dropped and another timeslice will be put at the end of the set.
+  std::deque<std::array<std::array<float, MaxPads>, MaxRows>> mSetOfTimeSlices{};
```
Comment (wiechula) [inline Detectors/TPC/reconstruction/include/TPCReconstruction/KrBoxClusterFinder.h]:
```suggestion
  std::deque<TimeSliceSector> mSetOfTimeSlices{};
```

File: Detectors/TPC/reconstruction/src/KrBoxClusterFinder.cxx [hunk 4]
```diff
@@ -112,7 +86,53 @@ void KrBoxClusterFinder::fillADCValue(int cru, int rowInSector, int padInRow, in
     adcValue /= correctionFactor;
   }
 
-  mMapOfAllDigits[timeBin][rowInSector][corPad] = adcValue;
+  mTempTimeSlice[rowInSector][corPad] = adcValue;
+}
+
+void KrBoxClusterFinder::fillTempTimeSlice(std::vector<o2::tpc::Digit>& eventSector, const int timeSlice) {
```
Comment (wiechula) [inline Detectors/TPC/reconstruction/src/KrBoxClusterFinder.cxx]: As it is implemented now, digits will be removed from the data. This means that devices after this one could not us the digits any longer. Also, this implies copy operations, most probably. Computing wise, it might be better to add a digits counter per sector in the header `std::array<size_t, 36> mFirstDigit;` and use:
```
for (mFirstDigit[mSector]; mFirstDigit[mSector] < eventSector.size(); ++mFirstDigit[mSector]) {
  const auto& digit = eventSector[mFirstDigit[mSector]];
...
```
`mFirstDigit` would then need to be reset for each TF, of course. Perhaps in `loopOverSector` `mFirstDigit[mSector] = 0;`. And `mSector` must be set properly before calling `loopOverSector`.
It it is clear that `loopOverSector` will only be called once per instance of `KrBoxClusterFinder`, then of course one could simply use `size_t mFirstDigit` and life would be simpler.

But we could also for the moment keep it like this.

File: Detectors/TPC/reconstruction/src/KrBoxClusterFinder.cxx (hunk 1 above)
Comment (wiechula) [reply Detectors/TPC/reconstruction/src/KrBoxClusterFinder.cxx]: I think we will need to implement it in a different way. I already have an idea.

File: Detectors/TPC/reconstruction/src/KrBoxClusterFinder.cxx [hunk 5]
```diff
@@ -112,7 +86,53 @@ void KrBoxClusterFinder::fillADCValue(int cru, int rowInSector, int padInRow, in
...
+  return;
+}
+
+const std::vector<KrCluster>& KrBoxClusterFinder::loopOverSector(std::vector<o2::tpc::Digit>& eventSector) {
+  mClusters.clear();
+  createInitialMap(eventSector);
+  for(int iTimeSlice = mMaxClusterSizeTime; iTimeSlice < MaxTimes - mMaxClusterSizeTime; ++iTimeSlice) {
+    findLocalMaxima(true, iTimeSlice);
+    popFirstTimeSliceFromMap();
+    fillTempTimeSlice(eventSector, iTimeSlice);
```
Comment (wiechula) [inline Detectors/TPC/reconstruction/src/KrBoxClusterFinder.cxx]: Why is this not
```suggestion
    fillTempTimeSlice(eventSector, iTimeSlice + mMaxClusterSizeTime + 1);
```
? I thought the new slice should be added at the end, or? The `createInitialMap` goes up to `2*mMaxClusterSizeTime` including the last.

**Review by wiechula — CHANGES_REQUESTED** [review]

**Comment by wiechula** [comment]
Two more general comments.
1. While it is simple, of course to treat one sector as `max rows * max pads` this brings a size overhead of a factor 1.44. We should think about avoiding this. Should be rather simple to implement.
2. I like the present implementation, since it is very memory efficient. Only time bins which are strictly required for the cluster finding are kept in memory. We sitll might need to change this in future once the final zero suppression logic is in place. Just to keep in mind.

**Comment by wiechula** [comment]
@HauenderPhilip , please fix the formatting and have a look at my questions / suggestions.

File: Detectors/TPC/reconstruction/src/KrBoxClusterFinder.cxx (hunk 5 above)
Comment (HauenderPhilip) [reply Detectors/TPC/reconstruction/src/KrBoxClusterFinder.cxx]: You are right, yes!
This bug, however, had no big impact for the cluster finding, since it only affected the first few timebins. Nevertheless, I fixed it now. Thanks!

File: Detectors/TPC/reconstruction/src/KrBoxClusterFinder.cxx (hunk 4 above)
Comment (HauenderPhilip) [reply Detectors/TPC/reconstruction/src/KrBoxClusterFinder.cxx]: I don't get the point, that this removes digits which can not be used afterwards. Is this not just a copy of digits?
The other points are clear and I implemented them. Thank you.

File: Detectors/TPC/reconstruction/src/KrBoxClusterFinder.cxx (hunk 1 above)
Comment (HauenderPhilip) [reply Detectors/TPC/reconstruction/src/KrBoxClusterFinder.cxx]: Very cool, thanks. Do you want to explain your ideas?

File: Detectors/TPC/reconstruction/src/KrBoxClusterFinder.cxx (hunk 4 above)
Comment (wiechula) [reply Detectors/TPC/reconstruction/src/KrBoxClusterFinder.cxx]: If we use the code in DPL we will receive the digits in the device. Since the function received the vector by reference, the vector will be empty at the end, since `loopOverSector` will loop over all time bins and `eventSector.erase(digit);` will remove the digit. So the `std::vector<o2::tpc::Digit>` passed from the DPL device to `loopOverSector` will then by empty in the DPL device.

File: Detectors/TPC/reconstruction/src/KrBoxClusterFinder.cxx (hunk 1 above)
Comment (wiechula) [reply Detectors/TPC/reconstruction/src/KrBoxClusterFinder.cxx]: The DPL device currently fills the data digit by digit. We will now first need to buffer all digits and call `loopOverSector` instead of `fillADCValue`. I can quickly implement it once the changes are committed.

File: Detectors/TPC/reconstruction/include/TPCReconstruction/KrBoxClusterFinder.h [hunk 6]
```diff
@@ -163,6 +170,17 @@ class KrBoxClusterFinder
...
+  void fillADCValue(int cru, int rowInSector, int padInRow, int timeBin, float adcValue);
+  void resetADCMap();
+
+  void createInitialMap(std::vector<o2::tpc::Digit>& eventSector);
+  void addTimeSliceToMap();
+  void popFirstTimeSliceFromMap();
+  void fillADCValueInTempSlice(int cru, int rowInSector, int padInRow, float adcValue);
+  void fillTempTimeSlice(std::vector<o2::tpc::Digit>& eventSector, const int timeSlice);
+  void clearTempTimeSlice();
+  const std::vector<KrCluster>& loopOverSector(std::vector<o2::tpc::Digit>& eventSector);
```
Comment (wiechula) [inline Detectors/TPC/reconstruction/include/TPCReconstruction/KrBoxClusterFinder.h]:
```suggestion
  void loopOverSector(std::vector<o2::tpc::Digit>& eventSector);
```
I think it doesn't make sense to return mClusters, since there is already a getter for this.

File: Detectors/TPC/reconstruction/include/TPCReconstruction/KrBoxClusterFinder.h [hunk 7]
```diff
@@ -187,35 +205,55 @@ class KrBoxClusterFinder
...
   KrCluster mTempCluster; ///< Used to save the cluster data
 
-  /// Here the map is defined where all digits are temporarily stored
-  std::array<std::array<std::array<float, MaxPads>, MaxRows>, MaxTimes> mMapOfAllDigits{};
+  using TimeSliceSector = std::array<std::array<float, MaxPads>, MaxRows>;
+
+  /// A temporary timeslice that gets filled and can be added to the set of timeslices
+  TimeSliceSector mTempTimeSlice;
+
+  std::array<size_t, 36> mFirstDigit = {0};
```
Comment (wiechula) [inline Detectors/TPC/reconstruction/include/TPCReconstruction/KrBoxClusterFinder.h]:
```suggestion
  size_t mFirstDigit = {0};
```
Since the KrBoxClusterFinder can only handle one sector, it does not make sense to create an array for all sectors here.

File: Detectors/TPC/reconstruction/src/KrBoxClusterFinder.cxx [hunk 8]
```diff
@@ -112,7 +90,59 @@ void KrBoxClusterFinder::fillADCValue(int cru, int rowInSector, int padInRow, in
...
+
+    // We don't need this digit anymore:
+    // digit = eventSector.erase(digit);
+  }
+  return;
+}
+
+const std::vector<KrCluster>& KrBoxClusterFinder::loopOverSector(std::vector<o2::tpc::Digit>& eventSector)
+{
+  mFirstDigit[mSector] = 0;
```
Comment (wiechula) [inline Detectors/TPC/reconstruction/src/KrBoxClusterFinder.cxx]:
```suggestion
  mFirstDigit = 0;
```

File: Detectors/TPC/reconstruction/src/KrBoxClusterFinder.cxx [hunk 9]
```diff
@@ -112,7 +90,59 @@ void KrBoxClusterFinder::fillADCValue(int cru, int rowInSector, int padInRow, in
...
 
-  mMapOfAllDigits[timeBin][rowInSector][corPad] = adcValue;
+  mTempTimeSlice[rowInSector][corPad] = adcValue;
+}
+
+void KrBoxClusterFinder::fillTempTimeSlice(std::vector<o2::tpc::Digit>& eventSector, const int timeSlice)
+{
+  // No third argument needed since iterator gets popped
+  for (mFirstDigit[mSector]; mFirstDigit[mSector] < eventSector.size(); ++mFirstDigit[mSector]) {
+    const auto& digit = eventSector[mFirstDigit[mSector]];
```
Comment (wiechula) [inline Detectors/TPC/reconstruction/src/KrBoxClusterFinder.cxx]:
```suggestion
  for (mFirstDigit; mFirstDigit < eventSector.size(); ++mFirstDigit) {
    const auto& digit = eventSector[mFirstDigit];
```

File: Detectors/TPC/reconstruction/include/TPCReconstruction/KrBoxClusterFinder.h [hunk 10]
```diff
@@ -47,23 +46,29 @@
...
+/// How to use (see macro: findKrBoxCluster.C):
+/// Create KrBoxClusterFinder object
+/// Load gainmap (if available)
+/// Make adjustments to clusterfinder (e.g. set min number of neighbours)
 /// Loop over all sectors
-/// Use KrBoxClusterFinder(sector) to create 3D map
-/// Use localMaxima() to find..well the local maxima
-/// localMaxima() yields a vector of a tuple with three int
-/// Loop over this vector and give the tuple to buildCluster()
+/// Use "loopOverSector" function for this
```
Comment (wiechula) [inline Detectors/TPC/reconstruction/include/TPCReconstruction/KrBoxClusterFinder.h:72]: Please document somewhere that also `setSector` has to be called and that processing can only be done sector after sector. We could think about getting the sector from the Digits.

File: Detectors/TPC/reconstruction/include/TPCReconstruction/KrBoxClusterFinder.h [hunk 11]
```diff
@@ -96,6 +102,13 @@ namespace tpc
 
 class KrBoxClusterFinder
 {
+ private:
+  /// Maximum Map Dimensions
+  /// Here is room for improvements
+  static constexpr size_t MaxPads = 138; ///< Size of the map in pad-direction
+  static constexpr size_t MaxRows = 152; ///< Size of the map in row-direction
+  size_t MaxTimes = 114048;              ///< Size of the map in time-direction
```
Comment (wiechula) [inline Detectors/TPC/reconstruction/include/TPCReconstruction/KrBoxClusterFinder.h]:
```suggestion
  size_t mMaxTimes = 114048;              ///< Size of the map in time-direction
```

File: Detectors/TPC/reconstruction/include/TPCReconstruction/KrBoxClusterFinder.h [hunk 12]
```diff
@@ -145,6 +149,9 @@ class KrBoxClusterFinder
   /// Set Function for minimal charge required for maxCharge of a cluster
   void setMinQTreshold(int minQThreshold) { mQThresholdMax = minQThreshold; }
 
+  /// Set Function for minimal charge required for maxCharge of a cluster
+  void setMaxTimes(int maxTimes) { MaxTimes = maxTimes; }
```
Comment (wiechula) [inline Detectors/TPC/reconstruction/include/TPCReconstruction/KrBoxClusterFinder.h]:
```suggestion
  void setMaxTimes(int maxTimes) { mMaxTimes = maxTimes; }
```

File: Detectors/TPC/reconstruction/include/TPCReconstruction/KrBoxClusterFinder.h [hunk 13]
```diff
@@ -163,6 +170,17 @@ class KrBoxClusterFinder
...
 
+  void fillADCValue(int cru, int rowInSector, int padInRow, int timeBin, float adcValue);
+  void resetADCMap();
+
+  void createInitialMap(std::vector<o2::tpc::Digit>& eventSector);
+  void addTimeSliceToMap();
+  void popFirstTimeSliceFromMap();
+  void fillADCValueInTempSlice(int cru, int rowInSector, int padInRow, float adcValue);
+  void fillTempTimeSlice(std::vector<o2::tpc::Digit>& eventSector, const int timeSlice);
+  void clearTempTimeSlice();
```
Comment (wiechula) [inline Detectors/TPC/reconstruction/include/TPCReconstruction/KrBoxClusterFinder.h]: Shouldn't they be private?

File: Detectors/TPC/reconstruction/macro/findKrBoxCluster.C [hunk 14]
```diff
@@ -58,44 +58,28 @@ void findKrBoxCluster(int lastTimeBin = 1000, int run = -1, int time = -1, std::
     clFinder->loadGainMapFromFile(gainMapFile);
   }
```
Comment (wiechula) [inline Detectors/TPC/reconstruction/macro/findKrBoxCluster.C:60]: I propose to remove
```c++
auto clFinder = std::make_unique<o2::tpc::KrBoxClusterFinder>();
```
here, and the above, when the tree is created use replace
```c++
std::vector<o2::tpc::KrCluster> clusters;
```
by
```c++
auto clFinder = std::make_unique<o2::tpc::KrBoxClusterFinder>();
auto& clusters = clFinder->getClusters();
```
This will avoid the unnecessary copy of the clusters.
See also comment below.

File: Detectors/TPC/reconstruction/macro/findKrBoxCluster.C [hunk 15]
```diff
@@ -58,44 +58,28 @@ void findKrBoxCluster(int lastTimeBin = 1000, int run = -1, int time = -1, std::
...
     tree->GetEntry(iEvent);
-    // Each event consists of sectors (atm only two)
 
     for (int i = 0; i < 36; i++) {
       auto sector = digitizedSignal[i];
       if (sector->size() == 0) {
         continue;
       }
+      clFinder->setSector(i);
+      clusters = clFinder->loopOverSector(*sector);
```
Comment (wiechula) [inline Detectors/TPC/reconstruction/macro/findKrBoxCluster.C]:
```suggestion
      clFinder->loopOverSector(*sector);
```
With my other comments.

File: Detectors/TPC/reconstruction/macro/findKrBoxCluster.C [hunk 16]
```diff
@@ -58,44 +58,28 @@ void findKrBoxCluster(int lastTimeBin = 1000, int run = -1, int time = -1, std::
...
-        }
-        // Build total cluster
-        o2::tpc::KrCluster tempCluster = clFinder->buildCluster(padMax, rowMax, timeMax);
-        tempCluster.sector = i;
-
-        clusters.emplace_back(tempCluster);
-      }
+      // Fill Tree
+      tClusters->Fill();
+      clusters.clear();
```
Comment (wiechula) [inline Detectors/TPC/reconstruction/macro/findKrBoxCluster.C]:
```suggestion
```
The clean will not be needed, since `clusters` is now a refrence to `mClusters` and the clear will be called in `loopOverSector`.

File: Detectors/TPC/reconstruction/src/KrBoxClusterFinder.cxx [hunk 17]
```diff
@@ -112,7 +90,59 @@ void KrBoxClusterFinder::fillADCValue(int cru, int rowInSector, int padInRow, in
     adcValue /= correctionFactor;
   }
 
-  mMapOfAllDigits[timeBin][rowInSector][corPad] = adcValue;
+  mTempTimeSlice[rowInSector][corPad] = adcValue;
+}
+
+void KrBoxClusterFinder::fillTempTimeSlice(std::vector<o2::tpc::Digit>& eventSector, const int timeSlice)
```
Comment (wiechula) [inline Detectors/TPC/reconstruction/src/KrBoxClusterFinder.cxx]: Can't one simplify the filling of the slice and get rid of the `mTempTimeSlice`, remove
```
  void addTimeSliceToMap();
  void clearTempTimeSlice();
```

And call this function
```suggestion
void KrBoxClusterFinder::addTimeSlice(std::vector<o2::tpc::Digit>& eventSector, const int timeSlice)
```
together with the canges below.

This will avoid copying and unnecessary complexity.

File: Detectors/TPC/reconstruction/src/KrBoxClusterFinder.cxx [hunk 18]
```diff
@@ -112,7 +90,59 @@ void KrBoxClusterFinder::fillADCValue(int cru, int rowInSector, int padInRow, in
...
+{
```
Comment (wiechula) [inline Detectors/TPC/reconstruction/src/KrBoxClusterFinder.cxx:85]: Add
```c++
mSetOfTimeSlices.emplace_back();
```

File: Detectors/TPC/reconstruction/src/KrBoxClusterFinder.cxx [hunk 19]
```diff
@@ -112,7 +90,59 @@ void KrBoxClusterFinder::fillADCValue(int cru, int rowInSector, int padInRow, in
...
+    mSector = cru / CRU::CRUperSector;
+
+    const int rowInSector = digit.getRow();
+    const int padInRow = digit.getPad();
+    float adcValue = digit.getChargeFloat();
+
+    fillADCValueInTempSlice(cru, rowInSector, padInRow, adcValue);
+
+    // We don't need this digit anymore:
+    // digit = eventSector.erase(digit);
```
Comment (wiechula) [inline Detectors/TPC/reconstruction/src/KrBoxClusterFinder.cxx]:
```suggestion
    fillADCValueInLastSlice(cru, rowInSector, padInRow, adcValue);
```

File: Detectors/TPC/reconstruction/src/KrBoxClusterFinder.cxx [hunk 20]
```diff
@@ -38,66 +37,45 @@ void KrBoxClusterFinder::loadGainMapFromFile(const std::string_view calDetFileNa
...
 
-  // Every row starts at pad zero. But the number of pads in a row is not a constant.
-  // If we would just fill the map naively, we would put pads next to each other, which are not neighbours on the pad plane.
-  // Hence, we need to correct for this:
-  mSector = cru / CRU::CRUperSector;
-  const int pads = mMapperInstance.getNumberOfPadsInRowSector(rowInSector);
-  const int corPad = padInRow - (pads / 2) + (MaxPads / 2);
+void KrBoxClusterFinder::fillADCValueInTempSlice(int cru, int rowInSector, int padInRow, float adcValue)
+{
+  // Correct for pad offset:
```
Comment (wiechula) [inline Detectors/TPC/reconstruction/src/KrBoxClusterFinder.cxx:59]: auto& timeSlice = mSetOfTimeSlices.back();

File: Detectors/TPC/reconstruction/src/KrBoxClusterFinder.cxx [hunk 21]
```diff
@@ -38,66 +37,45 @@ void KrBoxClusterFinder::loadGainMapFromFile(const std::string_view calDetFileNa
...
+  const int padsInRow = mMapperInstance.getNumberOfPadsInRowSector(rowInSector);
+  const int corPad = padInRow - (padsInRow / 2) + (MaxPads / 2);
 
+  // Get correction factor from gain map:
   const auto correctionFactorCalDet = mGainMap.get();
   if (!correctionFactorCalDet) {
-    mMapOfAllDigits[timeBin][rowInSector][corPad] = adcValue;
+    mTempTimeSlice[rowInSector][corPad] = adcValue;
```
Comment (wiechula) [inline Detectors/TPC/reconstruction/src/KrBoxClusterFinder.cxx]:
```suggestion
    timeSlice[rowInSector][corPad] = adcValue;
```

File: Detectors/TPC/reconstruction/src/KrBoxClusterFinder.cxx [hunk 22]
```diff
@@ -112,7 +90,59 @@ void KrBoxClusterFinder::fillADCValue(int cru, int rowInSector, int padInRow, in
     adcValue /= correctionFactor;
   }
 
-  mMapOfAllDigits[timeBin][rowInSector][corPad] = adcValue;
+  mTempTimeSlice[rowInSector][corPad] = adcValue;
```
Comment (wiechula) [inline Detectors/TPC/reconstruction/src/KrBoxClusterFinder.cxx]:
```suggestion
  timeSlice[rowInSector][corPad] = adcValue;
```

File: Detectors/TPC/reconstruction/src/KrBoxClusterFinder.cxx [hunk 23]
```diff
@@ -38,66 +37,45 @@ void KrBoxClusterFinder::loadGainMapFromFile(const std::string_view calDetFileNa
...
   }
+}
 
-  // Every row starts at pad zero. But the number of pads in a row is not a constant.
-  // If we would just fill the map naively, we would put pads next to each other, which are not neighbours on the pad plane.
-  // Hence, we need to correct for this:
-  mSector = cru / CRU::CRUperSector;
-  const int pads = mMapperInstance.getNumberOfPadsInRowSector(rowInSector);
-  const int corPad = padInRow - (pads / 2) + (MaxPads / 2);
+void KrBoxClusterFinder::fillADCValueInTempSlice(int cru, int rowInSector, int padInRow, float adcValue)
```
Comment (wiechula) [inline Detectors/TPC/reconstruction/src/KrBoxClusterFinder.cxx]:
```suggestion
void KrBoxClusterFinder::fillADCValueInLastSlice(int cru, int rowInSector, int padInRow, float adcValue)
```

File: Detectors/TPC/reconstruction/src/KrBoxClusterFinder.cxx [hunk 24]
```diff
@@ -112,7 +90,59 @@ void KrBoxClusterFinder::fillADCValue(int cru, int rowInSector, int padInRow, in
...
+{
+  mFirstDigit[mSector] = 0;
+  mClusters.clear();
+  createInitialMap(eventSector);
+  for (int iTimeSlice = mMaxClusterSizeTime; iTimeSlice < MaxTimes - mMaxClusterSizeTime; ++iTimeSlice) {
+    findLocalMaxima(true, iTimeSlice);
+    popFirstTimeSliceFromMap();
+    fillTempTimeSlice(eventSector, iTimeSlice + mMaxClusterSizeTime + 1);
+    addTimeSliceToMap();
+    clearTempTimeSlice();
```
Comment (wiechula) [inline Detectors/TPC/reconstruction/src/KrBoxClusterFinder.cxx]:
```suggestion
    addTimeSlice(eventSector, iTimeSlice + mMaxClusterSizeTime + 1);
```

File: Detectors/TPC/reconstruction/src/KrBoxClusterFinder.cxx [hunk 25]
```diff
@@ -38,66 +37,45 @@ void KrBoxClusterFinder::loadGainMapFromFile(const std::string_view calDetFileNa
...
-      //mMapOfAllDigits[iTime][iRow][iPad] = 0;
-      //}
-    }
+  mSetOfTimeSlices.clear();
+  clearTempTimeSlice();
+
+  for (int iTimeSlice = 0; iTimeSlice <= 2 * mMaxClusterSizeTime; ++iTimeSlice) {
+    fillTempTimeSlice(eventSector, iTimeSlice);
+    addTimeSliceToMap();
+    clearTempTimeSlice();
```
Comment (wiechula) [inline Detectors/TPC/reconstruction/src/KrBoxClusterFinder.cxx]:
```suggestion
    addTimeSlice(eventSector, iTimeSlice);
```

File: Detectors/TPC/reconstruction/src/KrBoxClusterFinder.cxx [hunk 26]
```diff
@@ -38,66 +37,45 @@ void KrBoxClusterFinder::loadGainMapFromFile(const std::string_view calDetFileNa
...
-  // Reset whole map:
-  for (int iTime = 0; iTime < MaxTimes; iTime++) {
-    for (int iRow = 0; iRow < MaxRows; iRow++) {
-      std::fill(mMapOfAllDigits[iTime][iRow].begin(), mMapOfAllDigits[iTime][iRow].end(), 0.);
-      //for (int iPad = 0; iPad < MaxPads; iPad++) {
-      //mMapOfAllDigits[iTime][iRow][iPad] = 0;
-      //}
-    }
+  mSetOfTimeSlices.clear();
+  clearTempTimeSlice();
```
Comment (wiechula) [inline Detectors/TPC/reconstruction/src/KrBoxClusterFinder.cxx]:
```suggestion
```

**Review by wiechula — CHANGES_REQUESTED** [review]

File: Detectors/TPC/reconstruction/src/KrBoxClusterFinder.cxx [hunk 27]
```diff
@@ -112,7 +77,56 @@ void KrBoxClusterFinder::fillADCValue(int cru, int rowInSector, int padInRow, in
...
+    const int cru = digit.getCRU();
+    mSector = cru / CRU::CRUperSector;
+
+    const int rowInSector = digit.getRow();
+    const int padInRow = digit.getPad();
+    float adcValue = digit.getChargeFloat();
+
+    fillADCValueInLastSlice(cru, rowInSector, padInRow, adcValue);
+  }
+  return;
```
Comment (wiechula) [inline Detectors/TPC/reconstruction/src/KrBoxClusterFinder.cxx]:
```suggestion
```

File: Detectors/TPC/reconstruction/src/KrBoxClusterFinder.cxx [hunk 28]
```diff
@@ -197,128 +213,134 @@ void KrBoxClusterFinder::updateTempCluster(float tempCharge, int tempPad, int te
...
-      } else if (iRow == MaxRowsIROC + MaxRowsOROC1) {
-        mMaxClusterSizePad = mMaxClusterSizePadOROC2;
-        mMaxClusterSizeRow = mMaxClusterSizeRowOROC2;
-      } else if (iRow == MaxRowsIROC + MaxRowsOROC1 + MaxRowsOROC2) {
-        mMaxClusterSizePad = mMaxClusterSizePadOROC3;
-        mMaxClusterSizeRow = mMaxClusterSizeRowOROC3;
-      }
+  // loop over whole mSetOfTimeSlices the find clusters
+  // for (int iTime = 0; iTime < mMaxTimes; iTime++) { //mSetOfTimeSlices.size()
+  int iTime = mMaxClusterSizeTime;
```
Comment (wiechula) [inline Detectors/TPC/reconstruction/src/KrBoxClusterFinder.cxx]:
```suggestion
  const int iTime = mMaxClusterSizeTime;
```

**Review by wiechula — CHANGES_REQUESTED** [review]
Hi @HauenderPhilip , if I see correctly, you didn't update the `findKrBoxCluster.C` macro, yet. Please do so. Otherwise only minor comments.

**Comment by HauenderPhilip** [comment]
> Hi @HauenderPhilip , if I see correctly, you didn't update the `findKrBoxCluster.C` macro, yet. Please do so. Otherwise only minor comments.

Hey @wiechula,
thanks a lot for your comments. Concerning the `findKrBoxCluster.C` macro: I changed it such that it now uses the `getClusters()` function. So no more unecessary copy operations have to be done. Or what am I missing?

**Comment by wiechula** [comment]
@HauenderPhilip , in the present version of the file https://github.com/AliceO2Group/AliceO2/pull/6532/files#diff-92dc58619d30fc296ddfe8ec8526f65fbc68f64746ffde2735f95403c7fd7d5a I still don't see that my suggestions, mainly
```c++
auto clFinder = std::make_unique<o2::tpc::KrBoxClusterFinder>();
auto& clusters = clFinder->getClusters();
tClusters->Branch("cls", &clusters);
tClusters->Branch("run", &run);
tClusters->Branch("time", &time);
```
were implemented.

There is still a copy, now done differently using
```c++
clusters = clFinder->getClusters();
```
This line should be removed.

**Comment by HauenderPhilip** [comment]
Ok, thanks for the explanation!
I see how it improves the code and I fixed it.

File: Detectors/TPC/reconstruction/macro/findKrBoxCluster.C [hunk 29]
```diff
@@ -52,50 +54,31 @@ void findKrBoxCluster(int lastTimeBin = 1000, int run = -1, int time = -1, std::
...
-        tempCluster.sector = i;
-
-        clusters.emplace_back(tempCluster);
-      }
+      // Fill Tree
+      tClusters->Fill();
     }
-    // Fill Tree
-    tClusters->Fill();
-    clusters.clear();
```
Comment (wiechula) [inline Detectors/TPC/reconstruction/macro/findKrBoxCluster.C]: I think there is one more conceptual problem. If you clear the clusters in `loopOverSector` one entry in the tree will be done per sector. So one looses the one-to-one relation to the initial tree. I think it would be better to remove the clear from the `loopOverSector`, and call it manually as before. So revert this change.

File: Detectors/TPC/reconstruction/src/KrBoxClusterFinder.cxx [hunk 30]
```diff
@@ -112,7 +77,55 @@ void KrBoxClusterFinder::fillADCValue(int cru, int rowInSector, int padInRow, in
...
+    fillADCValueInLastSlice(cru, rowInSector, padInRow, adcValue);
+  }
+}
+
+void KrBoxClusterFinder::loopOverSector(std::vector<o2::tpc::Digit>& eventSector, const int sector)
+{
+  mFirstDigit = 0;
+  mSector = sector;
+
+  mClusters.clear();
```
Comment (wiechula) [inline Detectors/TPC/reconstruction/src/KrBoxClusterFinder.cxx]:
```suggestion
```

**Review by wiechula — CHANGES_REQUESTED** [review]
@HauenderPhilip , sorry, one more thing ...

File: Detectors/TPC/reconstruction/macro/findKrBoxCluster.C [hunk 31]
```diff
@@ -52,50 +54,32 @@ void findKrBoxCluster(int lastTimeBin = 1000, int run = -1, int time = -1, std::
...
-        }
-        // Build total cluster
-        o2::tpc::KrCluster tempCluster = clFinder->buildCluster(padMax, rowMax, timeMax);
-        tempCluster.sector = i;
-
-        clusters.emplace_back(tempCluster);
-      }
+      // Fill Tree
+      tClusters->Fill();
+      clusters.clear();
```
Comment (wiechula) [inline Detectors/TPC/reconstruction/macro/findKrBoxCluster.C]: Sorry, this has to go outside the sector loop.

**Review by wiechula — CHANGES_REQUESTED** [review]

**Review by wiechula — DISMISSED** [review]

**Review by wiechula — APPROVED** [review]
